core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala

```diff
@@ -181,11 +185,20 @@ private[spark] class OutputCommitCoordinator(conf: SparkConf, isDriver: Boolean)
             }
         }
       case None =>
-        logDebug(s"Stage $stage has completed, so not allowing attempt number $attemptNumber of" +
-          s"partition $partition to commit")
+        logDebug(s"Stage $stage has completed, so not allowing" +
+          s" attempt number $attemptNumber of partition $partition to commit")
         false
     }
   }
+
+  private def attemptFailed(
+      stage: StageId,
+      partition: PartitionId,
+      attempt: TaskAttemptNumber): Boolean = synchronized {
+    stageStates.get(stage).exists { state =>
+      state.failures.get(partition).exists(_.contains(attempt))
```

Comment:
minor: the one place this is called, you've already looked up the state.  so you could take the `StageState` instead of the `StageId` (or even just inline the function entirely, though I think it may be a bit easier to understand as a helper method).